ARROW-4626: [Flight] Add application-defined metadata to DoGet/DoPut

lihalite opened the pull request:
Also covers [ARROW-4627](https://issues.apache.org/jira/browse/ARROW-4627).

This is quite an enormous change, if preferred, I can do my best to try and separate changes.

pitrou commented on format/Flight.proto:
```diff
@@ -296,7 +301,8 @@ message FlightData {
  */
-message PutResult {}
+message PutResult {
+  bytes data_app_metadata = 1;
```
Same here.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -47,17 +48,52 @@ def do_get(self, context, ticket):
+        table = pa.Table.from_arrays(data, names=['a'])
+        return flight.GeneratorStream(
+            table.schema,
+            MetadataFlightServer.number_batches(table))
```
Can simply write `self.number_batches(table)`.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -47,17 +48,52 @@ def do_get(self, context, ticket):
+        counter = 0
+        while True:
+            try:
+                batch, buf = reader.read_with_metadata()
```
`batch` is not validated at some point?

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -47,17 +48,52 @@ def do_get(self, context, ticket):
+                assert counter == client_counter
+                writer.write(struct.pack('<i', counter))
+                counter += 1
+            except StopIteration:
```
Raising `StopIteration` doesn't sound great for a regular function or method. Perhaps return `None` instead?

lihalite replied:
It is the same thing that `read_next_batch` does. Perhaps we should provide an iterator to wrap this method instead?

pitrou replied:
Ah, right. Let's keep it like this then.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -345,3 +385,49 @@ def test_token_auth_invalid():
+                server_idx, = struct.unpack('<i', metadata.to_pybytes())
+                assert idx == server_idx
+                idx += 1
+            except StopIteration:
```
Same remark wrt. `StopIteration`.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -97,9 +128,14 @@ cdef class Result:
+    """The type of a FlightDescriptor."""
+
     UNKNOWN = 0
+    """An unknown descriptor type."""
```
Docstrings intertwined with regular code like this won't be introspected.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -292,12 +331,136 @@ cdef class FlightInfo:
+        return pyarrow_wrap_batch(batch), metadata
+
+
+cdef class FlightMessageReader(_CRecordBatchReader, _ReadPandasOption):
```
So this is exactly identical to `FlightStreamReader`, except that the implementation casts its reader to `CFlightMessageReader*` instead of `CFlightStreamReader*`. Is there a way to reconcile this? Perhaps we can have a common C++ interface (with different implementations potentially)...

lihalite replied:
Good point, I'll try to wrangle some of the duplication here.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -292,12 +331,136 @@ cdef class FlightInfo:
+        else:
+            metadata = None
+
+        return pyarrow_wrap_batch(batch), metadata
```
I think it would be nicer to return a namedtuple, perhaps with fields named `batch` and `metadata` (or `app_metadata`).

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -424,31 +595,38 @@ cdef class FlightClient:
 
     def do_put(self, descriptor: FlightDescriptor, schema: Schema,
                options: FlightCallOptions = None):
         """Upload data to a flight."""
```
Document return type?

pitrou commented on docs/source/format/Flight.rst:
```diff
@@ -0,0 +1,106 @@
+.. specific language governing permissions and limitations
+.. under the License.
+
+Arrow Flight RPC
```
Thanks for the doc additions!

lihalite replied:
Of course! We have teams that wanted fuller docs, so hope this is a reasonable starting point.

lihalite commented on cpp/src/arrow/flight/client.h:
```diff
@@ -57,6 +57,33 @@ class ARROW_EXPORT FlightCallOptions {
+ public:
+  virtual Status WriteWithMetadata(const RecordBatch& batch,
+                                   std::shared_ptr<Buffer> app_metadata,
+                                   bool allow_64bit = false) = 0;
```
I think the two should be fully independent, especially as I don't want to block writes on reads (though that does raise the point that we should test that thoroughly!). Or perhaps are we worried that other RPC transports might not support fully bidirectional communication?

pitrou replied:
Ok, I was just asking :)

wesm commented:
I'll try to leave some comments on this when I can -- I have been heads down on ARROW-3144 and haven't been doing code reviews for the last week+

lihalite commented:
@wesm no worries, this is also going to need rebasing once your PR lands.

lihalite commented:
I am going to close this until #4047 is merged just to keep the PR queue smaller, as this will need a bunch of rebasing.

lihalite commented:
Reopening now that #4047 is merged.

This also does half of https://issues.apache.org/jira/browse/ARROW-5143, enabling integration testing of (non-nested) dictionaries in Flight.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -468,3 +530,80 @@ def test_location_invalid():
+        def block_read():
+            reader.read_next_batch()
+            event.set()
+            with pytest.raises(pa.ArrowIOError, match=".*Cancel.*"):
```
Exceptions raised in threads are swallowed (you'll just get a printout, but `pytest` won't notice). You can have a function set a flag in its caller or something.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -468,3 +530,80 @@ def test_location_invalid():
+            reader.read_next_batch()
+            event.set()
+            with pytest.raises(pa.ArrowIOError, match=".*Cancel.*"):
+                reader.read_next_batch()
```
I think this is incorrect. This line could be executed before the `reader.cancel()` line below. You need some more sophisticated synchronization (perhaps two events?).

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -263,13 +316,13 @@ def flight_server(server_base, *args, **kwargs):
+    server_process = multiprocessing.Process(target=_server_thread)
+    server_process.start()
 
     yield location
```
Should have a `try...finally` around this so that the process is always terminated.

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -158,10 +200,21 @@ def do_get(self, context, ticket):
+    def slow_stream():
+        data1 = [pa.array([-10, -5, 0, 5, 10], type=pa.int32())]
+        yield pa.Table.from_arrays(data1, names=['a'])
+        time.sleep(60)
```
Not sure such a long delay makes sense?

lihalite commented:
Thanks for the feedback! Rebased & made fixes. @pitrou yes, I'm waiting for a Java review too.

lihalite commented:
There looks to be some spurious CI failures, especially on Windows. I'll try one last time to get those to succeed...

pitrou commented:
MinGW builds are unfortunately broken currently.

lihalite commented:
Flight cancellation tests look flaky, disabling those in CI.

lihalite commented:
Okay, unit tests finally pass 😄 Just waiting on Java review, I believe.

wesm commented:
Thanks @lidavidm! I reached out to @jacques-n to see when he can spend some time reviewing the Java side of this

wesm commented on cpp/src/arrow/flight/client.cc:
```diff
@@ -132,29 +132,66 @@ class GrpcClientAuthReader : public ClientAuthReader {
+                     std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream,
+                     std::unique_ptr<GrpcStreamReader>* out);
+  std::shared_ptr<Schema> schema() const override;
+  Status ReadNext(std::shared_ptr<RecordBatch>* out) override;
```
Sorry to phone in from the peanut gallery, but what do you think about introducing a result struct like

```
struct FlightStreamChunk {
  std::shared_ptr<RecordBatch> data;
  std::shared_ptr<Buffer> app_metadata;
};
```

lihalite replied:
Sure, that's a little more convenient for callers. We can't change the signature of `ReadNext` here though, only `ReadWithMetadata`.

wesm replied:
In https://github.com/apache/arrow/pull/4483 I introduced an `Iterator<T>` concept -- it's not clear to me how important it is to conform to the `RecordBatchReader` interface (or we could also use virtual inheritance so that the flight data reader can by dynamically-casted to `Iterator<std::shared_ptr<RecordBatch>>`) cc @pitrou

lihalite replied:
Hmm, maybe it's not so important, so long as in Python it still exposes the utility methods people expect (we find it useful to do `do_get().read_pandas()` as that's by far the common case). If we could expose both iterator-of-record-batch and iterator-of-chunk that would be best (as not everyone will want the metadata).

wesm replied:
On the C++ side having a intended API endpoint would probably the easiest thing (since the metadata can just be ignored if the developer-user doesn't care), and we can make the Python API as ergonomic as desired (I presume -- it is our intent in fact -- that most consumers of the Python API are going to be sitting behind a front-end interface to pyarrow anyway)

lihalite replied:
Alright, I went ahead and changed it to use that interface; when the other PR is merged, we can then make it actually implement `Iterator`.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -236,54 +262,74 @@ public void onCompleted() {
+    private final StreamListener<PutResult> listener;
+
+    SetStreamObserver(FlightProducer.StreamListener<PutResult> listener) {
+      this.listener = listener;
```
I suggest doing a noop listener here if input is null rather than the if(listener != null) down below.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -311,17 +357,34 @@ public PutResult getResult() {
+    void putNext();
 
-    public void error(Throwable ex);
+    void putNext(byte[] appMetadata);
```
what do you think about using a ArrowBuf here instead? No strong opinion, just thinking that would match the rest of the off-heapness of Arrow.

lihalite replied:
That's more consistent with the rest of Arrow, and with the C++ implementation (which uses `arrow::Buffer`, not `uint8_t*`). We'll have to be careful with making it clear who owns the buffer at what point, though.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/FlightStream.java:
```diff
@@ -56,10 +67,12 @@
   private volatile Throwable ex;
   private volatile FlightDescriptor descriptor;
   private volatile Schema schema;
+  private volatile byte[] applicationMetadata = null;
```
Let's construct as a listener pattern instead. Assuming people only want the last one seems a bit specific to what must be your use case.

lihalite replied:
Okay. I initially thought it would make correlating data with the associated metadata hard (and in the protocol, they're sent together - the metadata is not out-of-band for the cases that FlightStream represents), but the Java APIs are not really designed around manipulating individual RecordBatches anyways, so a listener would make sense.

lihalite replied:
@jacques-n on second thought, I don't think a listener is appropriate - it complicates usage quite a bit, especially if you want to correlate a particular metadatum with a particular RecordBatch. The current code does guarantee that `applicationMetadata` is not updated except by a call to `next`, which is the same invariant as for the `VectorSchemaRoot` contained in `FlightStream`. So if you do want previous metadata messages, you can retain them yourself (as you would with a listener anyways).

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/NoOpFlightProducer.java:
```diff
@@ -45,8 +41,8 @@ public FlightInfo getFlightInfo(CallContext context,
   @Override
-  public Callable<PutResult> acceptPut(CallContext context,
-      FlightStream flightStream) {
+  public Runnable acceptPut(CallContext context,
```
does it make sense to do some kind of Cancellable type thing here?

lihalite replied:
Do you mean being able to cancel an ongoing put from the server-side? Yes, we should expose that as well.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/PutResult.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class PutResult {
+
+  private ByteBuffer applicationMetadata;
```
Per elsewhere, let's make this ArrowBuf (or ByteBuf)

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/example/integration/IntegrationTestClient.java:
```diff
@@ -94,11 +96,34 @@ private void run(String[] args) throws ParseException, IOException {
+            }
+
+            @Override
+            public void onCompleted() {
```
Shouldn't this test wait until onCompleted is called for all messages (to avoid flaky test issues)?

jacques-n commented on java/flight/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:
```diff
@@ -99,7 +97,19 @@ public void putStream() throws Exception {
 
       VectorSchemaRoot root = VectorSchemaRoot.of(iv);
-      ClientStreamListener listener = c.startPut(FlightDescriptor.path("hello"), root);
+      ClientStreamListener listener = c.startPut(FlightDescriptor.path("hello"), root, new StreamListener<PutResult>() {
```
Maybe we should have a default impl of StreamListener<PutResult>?

lihalite replied:
Done.

lihalite commented:
Side note, I will bump the Netty dependency version...I think this is the source of Netty failing to find methods at runtime (since we specified a Netty version that conflicted with gRPC's).

lihalite commented:
Integration tests are broken, will investigate...

lihalite commented:
Ok, all that should be left is rebasing on https://github.com/apache/arrow/pull/4553.

wesm commented:
There's still a flake looks like. Do you want @jacques-n to have another look before we merge?

lihalite commented:
That's what I get for forgetting flake8...

I would appreciate any further thoughts Jacques has on the metadata API in Java. I'm not convinced on the callback-based API, as it makes the control flow complicated, especially when mixed with the blocking API for reading data.

lihalite commented:
Ok, tried to fix the last Python tests...looks like gRPC error messages change text across versions (of course), so once https://github.com/apache/arrow/pull/4484 lands, let's convert those tests to instead provide/check the gRPC error code in some way.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/PutResult.java:
```diff
@@ -0,0 +1,87 @@
+    applicationMetadata = metadata;
+  }
+
+  /** Create a PutResult with application-specific metadata. */
```
probably good to state reference counting behavior in this kind of method.

lihalite commented:
Rebased & addressed the last bit of feedback.

jacques-n commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -235,54 +253,74 @@ public void onCompleted() {
+
+    private final CompletableFuture<Void> result;
+    private final BufferAllocator allocator;
+    private final StreamListener<PutResult> listener;
```
My ask was slightly different. I was proposing that it doesn't make sense to have a completeable future and a listener in the same class. I'm struggling with the pattern where you're doing a double call in the on* event handling calls.

lihalite replied:
Ah, sorry - to be clear, you'd like to essentially merge `PutObserver` and `SetStreamListener`? That sounds reasonable.

jacques-n replied:
It's weird to have listener > future > listener. Why not just stack the two listeners and the second one can do the future result (if it wants to go from listener > future)?

lihalite replied:
I guess the issue stems from trying to expose the result in two places; asynchronously via an optional listener given when starting DoPut, and through a future accessible on the listener used to write data during a DoPut. Maybe instead of the optional user-supplied listener defaulting to a no-op one, we can provide an easy way to supply one that converts the result to a future. That is, the current API looks like

```java
final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, new StreamListener<>() {
    // ....
    @Override
    public void onCompleted() {
    }
});
// call writer.putNext...
writer.getResult(); // implicitly, we have to complete this future, while providing results to the listener above
```
and there is duplication in receiving the events from the server, instead we could have

```java
final FutureListener listener = new FutureListener(); // {
    // optionally override onNext to get application metadata
// };
final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, listener);
// call writer.putNext()...
listener.getResult();
```

Is that cleaner? There's a bit more boilerplate, but we could get rid of that if we _require_ the listener to be a FutureListener, create a default instance if the client-supplied listener is null, and have `writer.getResult()` just pass through to `listener.getResult`()`.

jacques-n commented:
I think that makes sense. Providing a simple listener impl for returning
the items via sync methods seems like a cleaner approach.

On Sat, Jun 22, 2019, 6:02 AM lihalite <notifications@github.com> wrote:

> *@lihalite* commented on this pull request.
> ------------------------------
>
> In java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java
> <https://github.com/apache/arrow/pull/4282#discussion_r296445157>:
>
> >      return stream;
>    }
>
> -  private static class SetStreamObserver<T> implements StreamObserver<T> {
> -    private final SettableFuture<T> result = SettableFuture.create();
> -    private volatile T resultLocal;
> +  private static class SetStreamObserver implements StreamObserver<Flight.PutResult> {
> +
> +    private final CompletableFuture<Void> result;
> +    private final BufferAllocator allocator;
> +    private final StreamListener<PutResult> listener;
>
> I guess the issue stems from trying to expose the result in two places;
> asynchronously via an optional listener given when starting DoPut, and
> through a future accessible on the listener used to write data during a
> DoPut. Maybe instead of the optional user-supplied listener defaulting to a
> no-op one, we can provide an easy way to supply one that converts the
> result to a future. That is, the current API looks like
>
> final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, new StreamListener<>() {
>     // ....
>     @Override
>     public void onCompleted() {
>     }
> });// call writer.putNext...
> writer.getResult(); // implicitly, we have to complete this future, while providing results to the listener above
>
> and there is duplication in receiving the events from the server, instead
> we could have
>
> final FutureListener listener = new FutureListener(); // {
>     // optionally override onNext to get application metadata// };final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, listener);// call writer.putNext()...
> listener.getResult();
>
> Is that cleaner? There's a bit more boilerplate, but we could get rid of
> that if we *require* the listener to be a FutureListener, create a
> default instance if the client-supplied listener is null, and have
> writer.getResult() just pass through to listener.getResult()`.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/4282?email_source=notifications&email_token=AABMYNVQTHDVFLDBJ24CJ4TP3YPHNA5CNFSM4HL2TJIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4K6U7Y#discussion_r296445157>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABMYNQ7GT4NMWNCDWETN73P3YPHNANCNFSM4HL2TJIA>
> .
>

lihalite commented:
Rebased and implemented a sync metadata interface as per @jacques-n's suggestions. Example: https://github.com/apache/arrow/blob/03733a73158edb7ec4a83c2cbcc5c95d12c5f035/java/flight/src/test/java/org/apache/arrow/flight/TestApplicationMetadata.java#L117

jacques-n commented:
Thanks @lihalite! Lgtm. +1 on the java side.

wesm commented:
Build failing to due to lint issues and failure in the integration tests. I'll be happy to merge once the tests are green

pitrou commented:
AppVeyor build at https://ci.appveyor.com/project/pitrou/arrow/builds/25528229

pitrou commented:
btw, @lihalite you could enable Travis-CI and AppVeyor on your Github fork to get faster CI results (the Apache build queues have become heavily clogged these days).

lihalite commented:
Ah thanks for the advice @pitrou, let me see about that...

pitrou commented:
There are bunch of failures in `test_flight.py` on the manylinux1 build:
https://travis-ci.org/apache/arrow/jobs/550276194#L3977

You should probably be able to reproduce them by following the instructions from `python/manylinux1/README.md`. Be sure to pass `PYTHON_VERSION="3.6"` (or 3.7).

lihalite commented:
Hmm, I built the wheels, spun up a VM, and ran the tests, but couldn't reproduce...I kicked off another build on my clone: https://travis-ci.com/lihalite/arrow/builds/116830174

lihalite commented:
Looks like it succeeds, modulo the JPype issues: https://travis-ci.com/lihalite/arrow/builds/116855846 (I reverted the change to use subprocesses in Python tests)

I kicked off another AppVeyor build just to validate that: https://ci.appveyor.com/project/lihalite/arrow/builds/25537671

wesm commented:
Rebased again now that I've fixed ARROW-5728 (though we also have ARROW-5732, unrelated, now, breaking builds) so we can see what the latest CI says

pitrou commented:
Well, the integration tests fail on Java Flight, e.g.:
```
Command failed: java -Dio.netty.tryReflectionSetAccessible=true -cp /home/travis/build/apache/arrow/java/flight/target/arrow-flight-0.14.0-SNAPSHOT-jar-with-dependencies.jar org.apache.arrow.flight.example.integration.IntegrationTestClient -port 31338 -j /home/travis/build/apache/arrow/integration_temp/generated_primitive.json
With output:
--------------
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.TimeoutException
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.run(IntegrationTestClient.java:138)
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.main(IntegrationTestClient.java:69)
Caused by: java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.run(IntegrationTestClient.java:136)
	... 1 more

--------------
Traceback (most recent call last):
  File "integration_test.py", line 1326, in _compare_flight_implementations
    consumer.flight_request(producer.FLIGHT_PORT, json_path)
  File "integration_test.py", line 1436, in flight_request
    run_cmd(cmd)
  File "integration_test.py", line 98, in run_cmd
    raise e
  File "integration_test.py", line 90, in run_cmd
    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
  File "/home/travis/build/apache/arrow/cpp-toolchain/lib/python3.6/subprocess.py", line 336, in check_output
    **kwargs).stdout
  File "/home/travis/build/apache/arrow/cpp-toolchain/lib/python3.6/subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['java', '-Dio.netty.tryReflectionSetAccessible=true', '-cp', '/home/travis/build/apache/arrow/java/flight/target/arrow-flight-0.14.0-SNAPSHOT-jar-with-dependencies.jar', 'org.apache.arrow.flight.example.integration.IntegrationTestClient', '-port', '31338', '-j', '/home/travis/build/apache/arrow/integration_temp/generated_primitive.json']' returned non-zero exit status 1.
```
https://travis-ci.org/apache/arrow/jobs/550576927

pitrou commented:
Restarting the integration test job in case it's just the CI machine being very slow.

pitrou commented:
Still fails.

lihalite commented:
Just got in and will look, thanks @pitrou.

lihalite commented:
Personal Travis build: https://travis-ci.com/lihalite/arrow/builds/116967040

lihalite commented:
This passes, modulo OSX suddenly missing SIMD symbols.

pitrou commented:
If you rebase from master it should fix the macOS issues. Sorry for this, we've been busy firefighting CI issues lately.

lihalite commented:
No worries, should've rebased originally...

Personal build: https://travis-ci.com/lihalite/arrow/builds/116979112

pitrou commented:
I wonder why there's no AppVeyor build. Do you have one on your fork?

lihalite commented:
I manually triggered one: https://ci.appveyor.com/project/lihalite/arrow/builds/25560414

`appveyor.yml` is set not to trigger builds unless the HEAD commit touches certain files, if I'm reading it right...

pitrou commented:
Ah, indeed. Ideally this clause would have examined all files in the PR, not only in the last commit, though.

wesm commented:
Appveyor just doesn't trigger about 5-10% of the time in my experience, probably a bug on their end

wesm commented:
Here's the running Appveyor build: https://ci.appveyor.com/project/lihalite/arrow/builds/25560411. Will probably wait a bit for a couple more of the main builds to run

lihalite commented:
Looks like tests finally pass, except R on AppVeyor (which I just kicked again).

nealrichardson commented:
The R failure is unrelated: https://downforeveryoneorjustme.com/cloud.r-project.org

wesm commented:
Merging, thanks @lihalite and code reviewers!

lihalite commented:
Thanks everyone, this has been quite the journey!

wesm commented:
@lihalite would you like to start a section about Flight-specific improvements in 0.14 in https://docs.google.com/document/d/1ljkW5tBh7cDfPRg_z6YY-1cbXfXxpfpRUlSatbV8128/edit?usp=sharing?

lihalite commented:
Sure, I'll take a look when I get a chance.
